src/commands/upgrade.rs

```diff
@@ -124,22 +120,48 @@ async fn upgrade_bindings(config: &Config, metadata: &Metadata, dry_run: bool) -
             .find(|dep| dep.name == "cargo-component-bindings")
         else {
             log::debug!(
-                "Workspace package {} doesn't depend on cargo-component-bindings",
-                package.name
+                "workspace package `{name}` doesn't depend on cargo-component-bindings",
+                name = package.name
             );
             continue;
         };
 
-        if bindings_dep.req == self_version {
-            config.terminal().status(
-                "Skipping",
-                format!(
-                    "package `{}` as it already uses the current bindings crate version",
-                    package.name
-                ),
-            )?;
-            continue;
-        }
+        let s = bindings_dep.req.to_string();
+        let version = match s.strip_prefix('^').unwrap_or(&s).parse::<Version>() {
+            Ok(v) => {
+                if this_version.major == v.major
```

Comment:
I'm trying to think about whether this makes sense, but I keep getting stuck. I think it's because I don't actually understand what the relationship is between different versions of cargo-component and the bindings crate, and therefore what — if anything — it means for them to be "semver-compatible with each other".

E.g. if cargo-component and the bindings both start at version 1.3, and then...

- cargo component has a new 1.4 release, is it allowed to rely on behaviour in the bindings crate that is only introduced in its corresponding version 1.4?
- the bindings crate has a new 1.4 release, is it allowed to assume that cargo-component will do something before or after it is built that is only implemented in cargo-component in its corresponding version 1.4?

I think what I'm slowly getting at is that I'm not sure "mutual semver compatibility" is actually meaningful.

One way out could be to always publish both in lockstep, and require an exact match on major.minor.

But again, because I don't understand what the relationship is _supposed_ to be, I might just be talking nonsense.

Reply:
The bindings crate does very little (today, that is; previously it did the _actual_ bindings generation) other than include the bindings generated by `cargo-component` in the user's source file; to that end, they both need to agree on where the bindings are located.

I don't anticipate that relationship changing very often, but when it does, I think it would necessitate a bump of a major version (or minor when major is still 0).

Right now they publish with version bumps in lock-step, but only because our publish script isn't smart enough to bump only crates which have changed, and transitively, those that depend on them.

And also, since we're still at `0.y.z` of `cargo-component` itself, every release is pretty much a breaking change and they should probably be published lock-step until a 1.0.0.

Once there is a 1.0.0, though, I hope the bindings crate remains at 1.0.0 and only updates for major releases and any 1.y.z version of `cargo-component` can use it without having to warn users or bump the dependency in user's `Cargo.toml`s.

Reply:
Okay, I'm convinced. Thanks for the explanation! 😊